### PR TITLE
{Core} Refactor code for passing arguments

### PR DESCRIPTION
**Pull request by jiasli**
**Related command**
`az login`

**Description**
Refactor code for passing arguments.

Pave way for 

- https://github.com/Azure/azure-cli/pull/30301
- https://github.com/Azure/azure-cli/pull/30302

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 1]
```diff
@@ -938,7 +939,7 @@ def _transform_subscription_for_multiapi(s, s_dict):
             s_dict[_MANAGED_BY_TENANTS] = [{_TENANT_ID: t.tenant_id} for t in s.managed_by_tenants]
 
 
-def _create_identity_instance(cli_ctx, *args, **kwargs):
+def _create_identity_instance(cli_ctx, authority, tenant_id=None, client_id=None):
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:942]: Very limited `args` (actually only one - `authority`) and `kwargs` are used. It's better to be explicit.

https://peps.python.org/pep-0020/

> Explicit is better than implicit.

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 2]
```diff
@@ -413,7 +413,8 @@ def get_login_credentials(self, resource=None, client_id=None, subscription_id=N
             credential = self._create_credential(account, client_id=client_id)
             external_credentials = []
             for external_tenant in external_tenants:
-                external_credentials.append(self._create_credential(account, external_tenant, client_id=client_id))
+                external_credentials.append(self._create_credential(account, tenant_id=external_tenant,
+                                                                    client_id=client_id))
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:417]: `_create_credential()` defines `tenant_id` as a keyword argument:

https://github.com/Azure/azure-cli/blob/0f28626d61c9a43379e28794b1fcf1d58dee20a1/src/azure-cli-core/azure/cli/core/_profile.py#L659

Calling `_create_credential()` should follow the same pattern.

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 3]
```diff
@@ -665,7 +666,7 @@ def _create_credential(self, account, tenant_id=None, client_id=None):
         """
         user_type = account[_USER_ENTITY][_USER_TYPE]
         username_or_sp_id = account[_USER_ENTITY][_USER_NAME]
-        tenant_id = tenant_id if tenant_id else account[_TENANT_ID]
+        tenant_id = tenant_id or account[_TENANT_ID]
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:669]: https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not

> x or y: if x is true, then x, else y

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 4]
```diff
@@ -413,7 +413,7 @@ def get_login_credentials(self, resource=None, client_id=None, subscription_id=N
             credential = self._create_credential(account, client_id=client_id)
             external_credentials = []
             for external_tenant in external_tenants:
-                external_credentials.append(self._create_credential(account, external_tenant, client_id=client_id))
+                external_credentials.append(self._create_credential(account, tenant_id=external_tenant))
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:416]: `_create_credential()` defines `tenant_id` as a keyword argument:

https://github.com/Azure/azure-cli/blob/0f28626d61c9a43379e28794b1fcf1d58dee20a1/src/azure-cli-core/azure/cli/core/_profile.py#L659

Calling `_create_credential()` should follow the same pattern.

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 5]
```diff
@@ -938,7 +938,7 @@ def _transform_subscription_for_multiapi(s, s_dict):
             s_dict[_MANAGED_BY_TENANTS] = [{_TENANT_ID: t.tenant_id} for t in s.managed_by_tenants]
 
 
-def _create_identity_instance(cli_ctx, *args, **kwargs):
+def _create_identity_instance(cli_ctx, authority, tenant_id=None):
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:941]: Very limited `args` (actually only one - `authority`) and `kwargs` are used:

image

It's better to be explicit.

https://peps.python.org/pep-0020/

> Explicit is better than implicit.

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 6]
```diff
@@ -413,7 +413,7 @@ def get_login_credentials(self, resource=None, client_id=None, subscription_id=N
             credential = self._create_credential(account, client_id=client_id)
             external_credentials = []
             for external_tenant in external_tenants:
-                external_credentials.append(self._create_credential(account, external_tenant, client_id=client_id))
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:416]: `client_id` is unused.

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 7]
```diff
@@ -362,12 +362,10 @@ def logout_all(self):
         identity.logout_all_users()
         identity.logout_all_service_principal()
 
-    def get_login_credentials(self, resource=None, client_id=None, subscription_id=None, aux_subscriptions=None,
-                              aux_tenants=None):
+    def get_login_credentials(self, resource=None, subscription_id=None, aux_subscriptions=None, aux_tenants=None):
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:365]: `client_id` is unused.

**Review by bebound — DISMISSED** [review]

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 8]
```diff
@@ -655,18 +653,17 @@ def _try_parse_msi_account_name(account):
                 return parts[0], (None if len(parts) <= 1 else parts[1])
         return None, None
 
-    def _create_credential(self, account, tenant_id=None, client_id=None):
+    def _create_credential(self, account, tenant_id=None):
```
[the remark this answers is not in the document]
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/_profile.py:656]: There is no guarantee when https://github.com/Azure/azure-cli/pull/30302 will be merged. Better to keep the code clean for now.

[the remark this answers is not in the document]
Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/_profile.py:656]: If there's possibility that `client_id` is needed in the future, I agree with Hang that we can keep it for now. There's no harm to leave it as it is, right?

[the remark this answers is not in the document]
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/_profile.py]: Ok, changed as requested, but `client_id` will only be supported by `get_raw_token()`, not `get_login_credentials()`, so I am still removing it  from `get_login_credentials()`.

**Review by evelyn-ys — APPROVED** [review]

**Review by bebound — APPROVED** [review]

**Pull request merged**